contracts/lens/OpinionatedOracle.sol

```diff
@@ -0,0 +1,102 @@
+pragma solidity =0.7.6;
+
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/libraries/TickMath.sol';
+import '../libraries/WeightedOracleLibrary.sol';
+import '../libraries/OracleLibrary.sol';
+
+contract OpinionatedOracle {
+    IUniswapV3Factory factory;
+
+    enum ManipulationResistance {
+        Dangerous, // Spot price from beginning of the block
+        Weak, // 1 minute
+        Medium, // 30 minutes
+        Strong // 1 day
+    }
+
+    uint32[4] quotePeriods = [0, 60 seconds, 30 minutes, 1 days];
+
+    /// @param _factory Address of the UniswapV3 Factory used to look up token pools
+    constructor(address _factory) {
+        factory = IUniswapV3Factory(_factory);
+    }
+
+    /// @notice Returns a quote for a given token amount according to a desired price manipulation resistance
+    /// @notice across a specified list of fee tiers
+    /// @param baseToken Address of the input token for the quote
+    /// @param quoteToken Address of the output token for the quote
+    /// @param baseTokenAmount The amount of token to be converted
+    /// @param resistance The desired resistance against price manipulation: Dangerous, Weak, Medium, Strong
+    /// @param feeTiers An array of fee tiers to gather prices from for the quote
+    /// @return quoteTokenAmount The amount of quoteToken received for baseTokenAmount of baseToken
+    function quote(
+        address baseToken,
+        address quoteToken,
+        uint128 baseTokenAmount,
+        ManipulationResistance resistance,
+        uint24[] memory feeTiers
+    ) public view returns (uint256 quoteTokenAmount) {
+        require(feeTiers.length > 0, 'FT');
+
+        uint32 quotePeriod = quotePeriods[uint256(resistance)];
+        address pool;
+        int24 meanWeightedTick;
+
+        // If the quote period is more than 0, we can calculate the mean tick over an interval and weight it by
+        // each pool's mean liquidity over the same period
+        if (quotePeriod > 0) {
+            WeightedOracleLibrary.PeriodObservation[] memory poolObservations =
+                new WeightedOracleLibrary.PeriodObservation[](feeTiers.length);
+
+            for (uint256 i = 0; i < feeTiers.length; i++) {
+                pool = factory.getPool(baseToken, quoteToken, feeTiers[i]);
```

Comment:
calling getPool on the factory (sload + call) is a bunch more expensive than computing the address and calling consult on it optimistically, since you revert if the pool does not exist anyway.

if we hard code the tick spacings by fee tier, we don't have to call factory at all, but that's a bummer because it won't support new fee tiers unless the code is upgradeable or we store fee tier mappings in this contract